### PR TITLE
Fix include/exclude all operators for select fields

### PR DESCRIPTION
**Pull request by patrykgruszka**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ✔️ 
| New feature/enhancement? (use the a.x branch)      | ❌
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️ 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     |  

## Description
This PR completes the fix for the segment filter UI issue where "including all of" and "excluding all of" operators were incorrectly available for single-select fields (Country, Timezone, Region, Locale, etc.).
- Remove `INCLUDING_ALL` and `EXCLUDING_ALL` operators from 'select' field type in `OperatorListTrait.php`
- These operators are logically impossible for single-select fields and now only appear for multiselect fields
- Added handling for edge cases in `ComplexRelationValueFilterQueryBuilder.php` for company custom fields

---
#### Test 1: Built-in Single-Select Fields
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a segment 
3. Filter on country 
4. See that the including all/ excluding all is not available

#### Test 2: Custom Company Fields
  1. Navigate to Admin > Custom Fields
  2. Create a custom single-select field for Company:
     - Name: "Test Select Field"
     - Type: "Select"
     - Object: "Company"
     - Add some options (e.g., "Option 1", "Option 2", "Option 3")
  3. Create a segment and filter on your new company select field
  4. Verify that "including all of" / "excluding all of" operators are NOT available
  5. Create another custom multiselect field for Company with the same options
  6. Create a segment and filter on the multiselect field
  7. Verify that "including all of" / "excluding all of" operators ARE available for multiselect fields

### CODE REVIEW

File: app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php [hunk 1]
```diff
@@ -72,17 +72,36 @@ public function onNegativeOperators(SegmentOperatorQueryBuilderEvent $event): vo
...
+                $event->stopPropagation();
+
+                return;
+            }
+        }
+
         $leadsTableAlias = $event->getLeadsTableAlias();
 
+        // Treat EXCLUDING_ALL with a single value as notIn
+        $operator = $event->operatorIsOneOf(OperatorOptions::EXCLUDING_ALL) ? 'notIn' : $event->getFilter()->getOperator();
```
Comment (biozshock) [inline app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php:99]:
```suggestion
        $operator = $event->operatorIsOneOf(OperatorOptions::EXCLUDING_ALL) ? OperatorOptions::EXCLUDING_ANY : $event->getFilter()->getOperator();
```

File: app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php [hunk 2]
```diff
@@ -163,15 +182,34 @@ public function onDefaultOperators(SegmentOperatorQueryBuilderEvent $event): voi
...
+                $event->addExpression('1 = 0');
+                $event->stopPropagation();
+
+                return;
+            }
+        }
+
         $leadsTableAlias = $event->getLeadsTableAlias();
 
+        $operator = $event->operatorIsOneOf(OperatorOptions::INCLUDING_ALL) ? 'in' : $event->getFilter()->getOperator();
```
Comment (biozshock) [inline app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php:209]:
```suggestion
        $operator = $event->operatorIsOneOf(OperatorOptions::INCLUDING_ALL) ? OperatorOptions::INCLUDING_ANY : $event->getFilter()->getOperator();
```

File: app/bundles/LeadBundle/Form/Type/FilterType.php [hunk 3]
```diff
@@ -59,6 +60,15 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                 $operator = array_key_first($operators);
             }
 
+            // Keep legacy operators available for existing saved segments, but not for new filters.
+            $legacyOperators  = [OperatorOptions::INCLUDING_ALL, OperatorOptions::EXCLUDING_ALL];
+            $isLegacyOperator = null !== $operator && in_array($operator, $legacyOperators, true);
+
+            if ($isLegacyOperator && !in_array($operator, $operators, true)) {
+                $multiSelectOperators = $this->listModel->getOperatorsForFieldType('multiselect');
+                $operators += array_filter($multiSelectOperators, static fn ($v) => $v === $operator);
```
Comment (biozshock) [inline app/bundles/LeadBundle/Form/Type/FilterType.php]:
```suggestion
                $deprecatedOperatorTypes = $this->listModel->getOperatorsForFieldType([
                    'include' => $legacyOperators,
                ]);
                $operators += array_filter($deprecatedOperatorTypes, static fn ($v) => $v === $operator);
```

File: app/bundles/LeadBundle/Form/Type/FilterType.php [hunk 4]
```diff
@@ -59,6 +60,15 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                 $operator = array_key_first($operators);
             }
 
+            // Keep legacy operators available for existing saved segments, but not for new filters.
```
Comment (biozshock) [inline app/bundles/LeadBundle/Form/Type/FilterType.php:63]:
```suggestion
            // Keep legacy operators available for existing saved segments, but not for new filters.
            // @see https://github.com/mautic/mautic/pull/16012
```
If there is an agreed notation to "remove in version x.xx" i would add it here too.

**Review by biozshock — CHANGES_REQUESTED** [review]
Please check suggestions and apply ones that you think are relevant.

File: app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php (hunk 1 above)
Comment (patrykgruszka) [reply app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php]: The test fails after this change, as `public const EXCLUDING_ANY = '!in';` which is not valid Doctrine method.

```
There was 1 error:

1) Mautic\LeadBundle\Tests\Command\UpdateLeadListCommandFunctionalTest::testCustomFieldSelectIncludeExclude with data set "exclude all with match" ('!in_all', 0, 1, [1])
Error: Call to undefined method Mautic\LeadBundle\Segment\Query\Expression\ExpressionBuilder::!in()

/var/www/html/app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php:104
/var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:246
/var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:206
/var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:56
/var/www/html/app/bundles/LeadBundle/Segment/Query/Filter/BaseFilterQueryBuilder.php:42
/var/www/html/app/bundles/LeadBundle/Segment/ContactSegmentFilter.php:155
/var/www/html/app/bundles/LeadBundle/Segment/Query/ContactSegmentQueryBuilder.php:78
/var/www/html/app/bundles/LeadBundle/Segment/ContactSegmentService.php:200
/var/www/html/app/bundles/LeadBundle/Segment/ContactSegmentService.php:42
/var/www/html/app/bundles/LeadBundle/Model/ListModel.php:377
/var/www/html/app/bundles/LeadBundle/Command/UpdateLeadListsCommand.php:245
/var/www/html/app/bundles/LeadBundle/Command/UpdateLeadListsCommand.php:153
/var/www/html/vendor/symfony/console/Command/Command.php:341
/var/www/html/vendor/symfony/console/Command/LazyCommand.php:77
/var/www/html/vendor/symfony/console/Tester/CommandTester.php:75
/var/www/html/app/bundles/CoreBundle/Test/AbstractMauticTestCase.php:206
/var/www/html/app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php:440
```

File: app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php (hunk 2 above)
Comment (patrykgruszka) [reply app/bundles/LeadBundle/EventListener/SegmentOperatorQuerySubscriber.php]: This will work, as `OperatorOptions::INCLUDING_ANY` is the same as `in`, but I wouldn't change it to keep it consistent with `notIn` variant.

**Comment by patrykgruszka** [comment]
@biozshock there was two more problems:

- Same `Dunno how to handle operator "in_all"` error for Company single select field, now fixed in: https://github.com/mautic/mautic/pull/16012/changes/49d51ff6335a4b3a2b6df3330bea622ae9c91df0.
- Using “including any of” with a multi select field of a company behaves like a “including all of” fixed in: https://github.com/mautic/mautic/pull/16012/commits/9f7729304bc4b3e00beace9697f446d4fa417209

I've also added your test and applied one suggestion. Could you have a look again, if everything make sense now?

File: app/bundles/LeadBundle/Segment/Query/Filter/ComplexRelationValueFilterQueryBuilder.php [hunk 5]
```diff
@@ -110,13 +111,76 @@ public function applyQuery(QueryBuilder $queryBuilder, ContactSegmentFilter $fil
...
+                    $expression = $queryBuilder->expr()->$filterGlue(...$expressions);
+                }
+                break;
+            case OperatorOptions::INCLUDING_ALL:
+                // Single-select field can't match all values at once - always false for multiple values.
+                if (is_array($filterParametersHolder) && count($filterParametersHolder) > 1) {
+                    $expression = $queryBuilder->expr()->and('1 = 0');
+                    break;
+                }
+                $expression = $queryBuilder->expr()->in(
```
Comment (biozshock) [inline app/bundles/LeadBundle/Segment/Query/Filter/ComplexRelationValueFilterQueryBuilder.php]: If there is a single value it's better to use an equals operator.

File: app/bundles/LeadBundle/Segment/Query/Filter/ComplexRelationValueFilterQueryBuilder.php [hunk 6]
```diff
@@ -110,13 +111,76 @@ public function applyQuery(QueryBuilder $queryBuilder, ContactSegmentFilter $fil
...
+                break;
+            case OperatorOptions::EXCLUDING_ALL:
+                // Single-select field can't hold all values at once - always true for multiple values.
+                if (is_array($filterParametersHolder) && count($filterParametersHolder) > 1) {
+                    $expression = $queryBuilder->expr()->and('1 = 1');
+                    break;
+                }
+                $expression = $queryBuilder->expr()->or(
+                    $queryBuilder->expr()->isNull($tableAlias.'.'.$filter->getField()),
+                    $queryBuilder->expr()->notIn($tableAlias.'.'.$filter->getField(), $filterParametersHolder)
```
Comment (biozshock) [inline app/bundles/LeadBundle/Segment/Query/Filter/ComplexRelationValueFilterQueryBuilder.php]: If there is a single value it's better to use a not equals operator.

File: app/bundles/LeadBundle/Segment/Query/Filter/ComplexRelationValueFilterQueryBuilder.php [hunk 7]
```diff
@@ -110,13 +111,76 @@ public function applyQuery(QueryBuilder $queryBuilder, ContactSegmentFilter $fil
...
             case '!multiselect':
-                $operator    = 'multiselect' === $filterOperator ? 'regexp' : 'notRegexp';
+                /**
+                 * Multi-select filter behavior:
+                 * - INCLUDING_ALL => AND regexp(...)
+                 * - INCLUDING_ANY => OR regexp(...)
+                 * - EXCLUDING_ALL => NOT(AND regexp(...)) OR field IS NULL
+                 * - EXCLUDING_ANY => AND notRegexp(...) OR field IS NULL
+                 */
+                $filterArray      = $filter->contactSegmentFilterCrate->getArray();
```
Comment (biozshock) [inline app/bundles/LeadBundle/Segment/Query/Filter/ComplexRelationValueFilterQueryBuilder.php:121]: Newly added code is not covered with integration (functional) test.
Can you please add one to the app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php ?

**Review by biozshock — CHANGES_REQUESTED** [review]

File: app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php [hunk 8]
```diff
@@ -337,6 +338,320 @@ public function testCustomFieldIncludeExclude(string $filter, $expected, array $
         );
     }
 
+    /**
+     * @param int|string $expected
+     * @param array<int> $addFieldsToSegment
+     */
+    #[DataProvider('provideSingleIncludeExclude')]
+    public function testCustomFieldSelectIncludeExclude(string $filter, $expected, int $addFieldToContact, array $addFieldsToSegment): void
```
Comment (biozshock) [inline app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php]: You can use the `int $expected`. I've missed the return of `getLeadCount` was an integer in 7.0.

File: app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php [hunk 9]
```diff
@@ -337,6 +338,320 @@ public function testCustomFieldIncludeExclude(string $filter, $expected, array $
...
+        yield 'exclude all with match' => [OperatorOptions::EXCLUDING_ALL, 0, 1, [1]];
+        yield 'exclude all with match multiple' => [OperatorOptions::EXCLUDING_ALL, 1, 1, [1, 2]]; // Multiple values always match "!in_all" with single value
+    }
+
+    /**
+     * @param int|string $expected
+     * @param array<int> $addFieldsToSegment
+     */
+    #[DataProvider('provideSingleIncludeExclude')]
+    public function testCompanyCustomFieldSelectIncludeExclude(string $filter, $expected, int $addFieldToCompany, array $addFieldsToSegment): void
```
Comment (biozshock) [inline app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php]: You can use the `int $expected`. I've missed the return of `getLeadCount` was an integer in 7.0.

File: app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php [hunk 10]
```diff
@@ -337,6 +338,320 @@ public function testCustomFieldIncludeExclude(string $filter, $expected, array $
...
+        Assert::assertSame($expected, $leadListRepository->getLeadCount([$segment->getId()]));
+    }
+
+    /**
+     * @param int|string $expected
+     * @param array<int> $addFieldsToCompany
+     * @param array<int> $addFieldsToSegment
+     */
+    #[DataProvider('provideIncludeExclude')]
+    public function testCompanyCustomFieldIncludeExclude(string $filter, $expected, array $addFieldsToCompany, array $addFieldsToSegment): void
```
Comment (biozshock) [inline app/bundles/LeadBundle/Tests/Command/UpdateLeadListCommandFunctionalTest.php]: You can use the `int $expected`. I've missed the return of `getLeadCount` was an integer in 7.0.

**Review by biozshock — CHANGES_REQUESTED** [review]

**Review by biozshock — APPROVED** [review]
Code looks good. Tested the functionality, looks good too.

**Comment by matbcvo** [comment]
@patrykgruszka, please rebase onto the 7.1 branch.

**Comment by patrykgruszka** [comment]
@matbcvo rebased on `7.1`

**Review by JonasLudwig1998 — APPROVED** [review]
Works as expected

**Review by matbcvo — APPROVED** [review]

**Comment by promptless-for-oss** [comment]
Promptless prepared a documentation update related to this change.

**Triggered by:** [mautic/mautic#16012](https://github.com/mautic/mautic/pull/16012)

Updated the segment filter documentation to clarify that "Including all of" and "Excluding all of" operators are only available for multiselect fields. Single-select fields (Country, Timezone, Region, Locale, custom select fields) support other operators including "Including any of" and "Excluding any of".

**Review:** [Update segment filter operator documentation](https://app.gopromptless.ai/suggestions/8d8a3b95-0853-4fec-b478-3a942e5855a4)

**Comment by adiati98** [comment]
Hi @patrykgruszka,

Promptless created a PR based on the changes in this PR. Can you please review the contents and see if they're align with your changes? TIA! ✨ 

https://github.com/mautic/user-documentation/pull/765
